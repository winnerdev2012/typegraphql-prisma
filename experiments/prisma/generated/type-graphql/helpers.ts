
export function transformFields(fields: Record<string, any>): Record<string, any> {
  return Object.fromEntries(
    Object.entries(fields)
      // remove __typename and others
      .filter(([key, value]) => !key.startsWith("__"))
      .map<[string, any]>(([key, value]) => {
        if (Object.keys(value).length === 0) {
          return [key, true];
        }
        return [key, transformFields(value)];
      }),
  );
}

