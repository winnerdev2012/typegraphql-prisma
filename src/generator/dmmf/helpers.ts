export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\()+(.+)+(\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\()+(.+)+(\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\()+([A-Za-z])+\:+(.+)+(?:\))+/;

export function parseDocumentationAttributes<TData extends object = object>(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
): Partial<TData> {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return {};
  }
  const rawAttributeArgs = attribute
    ?.match(attributeArgsRegex)?.[0]
    ?.slice(1, -1);
  const splitRawArgsArray = rawAttributeArgs?.split(",").map(it => it.trim());
  const parsedAttributeArgs =
    splitRawArgsArray &&
    (Object.fromEntries(
      splitRawArgsArray.map(it => {
        const [key, value] = it.split(": ");
        return [key, JSON.parse(value)];
      }),
    ) as Partial<TData>);
  return parsedAttributeArgs ?? {};
}
