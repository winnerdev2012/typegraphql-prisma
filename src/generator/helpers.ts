import { DMMF } from "@prisma/client/runtime";

import { DMMFTypeInfo } from "./types";
import { ModelKeys } from "./config";

export function noop() {}

export function getBaseModelTypeName(modelName: string) {
  // TODO: add proper support for swapping model types with custom types
  // return `Base${modelName}`;
  return modelName;
}

export function getFieldTSType(typeInfo: DMMFTypeInfo, modelNames: string[]) {
  let TSType: string;
  if (typeInfo.kind === "scalar") {
    TSType = mapScalarToTSType(typeInfo.type);
  } else if (typeInfo.kind === "object") {
    if (modelNames.includes(typeInfo.type)) {
      TSType = getBaseModelTypeName(typeInfo.type);
    } else {
      TSType = typeInfo.type;
    }
  } else if (typeInfo.kind === "enum") {
    TSType = `keyof typeof ${typeInfo.type}`;
  } else {
    throw new Error(`Unsupported field type kind: ${typeInfo.kind}`);
  }
  if (typeInfo.isList) {
    TSType += "[]";
  }
  if (!typeInfo.isRequired) {
    TSType += " | null";
  }
  return TSType;
}

export function mapScalarToTSType(scalar: string) {
  switch (scalar) {
    case "ID": {
      // TODO: detect proper type of id field
      return "string";
    }
    case "String": {
      return "string";
    }
    case "Boolean": {
      return "boolean";
    }
    case "DateTime": {
      return "Date";
    }
    case "Int":
    case "Float": {
      return "number";
    }
    default:
      throw new Error(`Unrecognized scalar type: ${scalar}`);
  }
}

export function getTypeGraphQLType(
  typeInfo: DMMFTypeInfo,
  modelNames: string[],
) {
  let GraphQLType: string;
  if (typeInfo.kind === "scalar") {
    GraphQLType = mapScalarToTypeGraphQLType(typeInfo.type);
  } else if (typeInfo.kind === "object") {
    if (modelNames.includes(typeInfo.type)) {
      GraphQLType = getBaseModelTypeName(typeInfo.type);
    } else {
      GraphQLType = typeInfo.type;
    }
  } else {
    GraphQLType = typeInfo.type;
  }
  if (typeInfo.isList) {
    GraphQLType = `[${GraphQLType}]`;
  }
  return GraphQLType;
}

export function mapScalarToTypeGraphQLType(scalar: string) {
  switch (scalar) {
    case "DateTime": {
      return "Date";
    }
    case "Boolean":
    case "String":
    case "ID":
    case "Int":
    case "Float": {
      return scalar;
    }
    default: {
      throw new Error(`Unrecognized scalar type: ${scalar}`);
    }
  }
}

export function selectInputTypeFromTypes(
  inputTypes: DMMF.SchemaArgInputType[],
): DMMF.SchemaArgInputType {
  return (
    inputTypes.find(it => it.kind === "object") ||
    inputTypes.find(it => it.kind === "enum") ||
    inputTypes[0]
  );
}

export function camelCase(str: string) {
  return str[0].toLowerCase() + str.slice(1);
}

export function pascalCase(str: string): string {
  return str[0].toUpperCase() + str.slice(1);
}

export function getMappedActionName(
  actionName: ModelKeys,
  methodName: string,
  mapping: DMMF.Mapping,
): string {
  switch (actionName) {
    case "findOne": {
      return camelCase(mapping.model);
    }
    case "findMany": {
      return camelCase(mapping.plural);
    }
    default: {
      return methodName;
    }
  }
}
