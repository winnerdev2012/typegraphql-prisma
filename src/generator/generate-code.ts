import { DMMF } from "@prisma/client/runtime";
import { Project } from "ts-morph";
import path from "path";

import { noop } from "./helpers";
import generateEnumFromDef from "./enum";
import generateObjectTypeClassFromModel from "./object-type-class";
import generateRelationsResolverClassesFromModel from "./resolvers/relations";
import {
  generateOutputTypeClassFromType,
  generateInputTypeClassFromType,
} from "./type-class";
import generateCrudResolverClassFromMapping from "./resolvers/full-crud";
import {
  resolversFolderName,
  relationsResolversFolderName,
  crudResolversFolderName,
  inputsFolderName,
  outputsFolderName,
  enumsFolderName,
  modelsFolderName,
} from "./config";
import {
  generateResolversBarrelFile,
  generateInputsBarrelFile,
  generateOutputsBarrelFile,
  generateIndexFile,
  generateModelsBarrelFile,
  generateEnumsBarrelFile,
} from "./imports";
import saveSourceFile from "../utils/saveSourceFile";
import { GenerateCodeOptions } from "./options";

export default async function generateCode(
  dmmf: DMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop,
) {
  const baseDirPath = options.outputDirPath;
  const project = new Project();
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map(model => model.name);

  log("Generating enums...");
  const datamodelEnumNames = dmmf.datamodel.enums.map(enumDef => enumDef.name);
  await Promise.all(
    dmmf.datamodel.enums.map(enumDef =>
      generateEnumFromDef(project, baseDirPath, enumDef),
    ),
  );
  await Promise.all(
    dmmf.schema.enums
      // skip enums from datamodel
      .filter(enumDef => !datamodelEnumNames.includes(enumDef.name))
      .map(enumDef => generateEnumFromDef(project, baseDirPath, enumDef)),
  );
  const emittedEnumNames = [
    ...new Set([
      ...dmmf.schema.enums.map(it => it.name),
      ...dmmf.datamodel.enums.map(it => it.name),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, "index.ts"),
    undefined,
    { overwrite: true },
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);
  await saveSourceFile(enumsBarrelExportSourceFile);

  log("Generating models...");
  await Promise.all(
    dmmf.datamodel.models.map(model =>
      generateObjectTypeClassFromModel(project, baseDirPath, model, modelNames),
    ),
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, "index.ts"),
    undefined,
    { overwrite: true },
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmf.datamodel.models.map(it => it.name),
  );
  await saveSourceFile(modelsBarrelExportSourceFile);

  log("Generating output types...");
  const rootTypes = dmmf.schema.outputTypes.filter(type =>
    ["Query", "Mutation"].includes(type.name),
  );
  const outputTypesToGenerate = dmmf.schema.outputTypes.filter(
    // skip generating models and root resolvers
    type => !modelNames.includes(type.name) && !rootTypes.includes(type),
  );
  await Promise.all(
    outputTypesToGenerate.map(type =>
      generateOutputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        modelNames,
      ),
    ),
  );
  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map(it => it.name),
  );
  await saveSourceFile(outputsBarrelExportSourceFile);

  log("Generating input types...");
  await Promise.all(
    dmmf.schema.inputTypes.map(type =>
      generateInputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        modelNames,
      ),
    ),
  );
  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmf.schema.inputTypes.map(it => it.name),
  );
  await saveSourceFile(inputsBarrelExportSourceFile);

  log("Generating relation resolvers...");
  const relationResolversData = await Promise.all(
    dmmf.datamodel.models
      .filter(model => model.fields.some(field => field.relationName))
      .map(model => {
        const outputType = dmmf.schema.outputTypes.find(
          type => type.name === model.name,
        )!;
        const mapping = dmmf.mappings.find(it => it.model === model.name)!;
        return generateRelationsResolverClassesFromModel(
          project,
          baseDirPath,
          model,
          mapping,
          outputType,
          modelNames,
        );
      }),
  );
  if (relationResolversData.length > 0) {
    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        "index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateResolversBarrelFile(
      "relations",
      relationResolversBarrelExportSourceFile,
      relationResolversData,
    );
    await saveSourceFile(relationResolversBarrelExportSourceFile);
  }

  log("Generating crud resolvers...");
  const crudResolversData = await Promise.all(
    dmmf.mappings.map(mapping =>
      generateCrudResolverClassFromMapping(
        project,
        baseDirPath,
        mapping,
        rootTypes,
        modelNames,
        options,
      ),
    ),
  );
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateResolversBarrelFile(
    "crud",
    crudResolversBarrelExportSourceFile,
    crudResolversData,
  );
  await saveSourceFile(crudResolversBarrelExportSourceFile);

  log("Generating index file");
  const indexSourceFile = project.createSourceFile(
    baseDirPath + "/index.ts",
    undefined,
    { overwrite: true },
  );
  generateIndexFile(indexSourceFile, relationResolversData.length > 0);
  await saveSourceFile(indexSourceFile);
}
