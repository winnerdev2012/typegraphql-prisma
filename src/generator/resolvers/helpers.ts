import { getFieldTSType, getTypeGraphQLType } from "../helpers";
import { DmmfDocument } from "../dmmf/dmmf-document";
import { DMMF } from "../dmmf/types";

export function generateCrudResolverClassMethodDeclaration(
  action: DMMF.Action,
  typeName: string,
  method: DMMF.SchemaField,
  argsTypeName: string | undefined,
  collectionName: string,
  dmmfDocument: DmmfDocument,
  mapping: DMMF.Mapping,
) {
  const returnTSType = getFieldTSType(
    method.outputType,
    dmmfDocument,
    false,
    mapping.model,
    typeName,
  );

  return {
    name: action.name,
    isAsync: true,
    returnType: `Promise<${returnTSType}>`,
    decorators: [
      {
        name: `TypeGraphQL.${action.operation}`,
        arguments: [
          `_returns => ${getTypeGraphQLType(
            method.outputType,
            dmmfDocument,
            mapping.model,
            typeName,
          )}`,
          `{
            nullable: ${!method.outputType.isRequired},
            description: undefined
          }`,
        ],
      },
    ],
    parameters: [
      ...(action.kind === "aggregate"
        ? []
        : [
            {
              name: "ctx",
              // TODO: import custom `ContextType`
              type: "any",
              decorators: [{ name: "TypeGraphQL.Ctx", arguments: [] }],
            },
          ]),
      ...(!argsTypeName
        ? []
        : [
            {
              name: "args",
              type: argsTypeName,
              decorators: [{ name: "TypeGraphQL.Args", arguments: [] }],
            },
          ]),
    ],
    statements:
      action.kind === "aggregate"
        ? [
            // it will expose field resolvers automatically
            `return new ${returnTSType}();`,
          ]
        : [
            `return ctx.prisma.${collectionName}.${action.kind}(${
              argsTypeName ? "args" : ""
            });`,
          ],
  };
}
