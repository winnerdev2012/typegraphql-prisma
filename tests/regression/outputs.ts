import { promises as fs } from "fs";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { generateCodeFromSchema } from "../helpers/generate-code";
import createReadGeneratedFile, {
  ReadGeneratedFile,
} from "../helpers/read-file";

describe("outputs", () => {
  let outputDirPath: string;
  let readGeneratedFile: ReadGeneratedFile;

  beforeEach(async () => {
    outputDirPath = generateArtifactsDirPath("regression-outputs");
    await fs.mkdir(outputDirPath, { recursive: true });
    readGeneratedFile = createReadGeneratedFile(outputDirPath);
  });

  it("should properly generate output type classes", async () => {
    const schema = /* prisma */ `
      model Sample {
        intIdField    Int       @id @default(autoincrement())
        stringField   String
        floatField    Float
        booleanField  Boolean
        dateField     DateTime
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const aggregateSampleTSFile = await readGeneratedFile(
      "/resolvers/outputs/AggregateSample.ts",
    );
    const batchPayloadTSFile = await readGeneratedFile(
      "/resolvers/outputs/BatchPayload.ts",
    );

    expect(aggregateSampleTSFile).toMatchSnapshot("AggregateSample");
    expect(batchPayloadTSFile).toMatchSnapshot("BatchPayload");
  });

  it("should properly generate args type classes for aggregate", async () => {
    const schema = /* prisma */ `
      model Sample {
        intIdField    Int       @id @default(autoincrement())
        stringField   String
        floatField    Float
        booleanField  Boolean
        dateField     DateTime
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const aggregateSampleTSFile = await readGeneratedFile(
      "/resolvers/outputs/args/AggregateSampleCountArgs.ts",
    );

    expect(aggregateSampleTSFile).toMatchSnapshot("AggregateSampleCountArgs");
  });
});
