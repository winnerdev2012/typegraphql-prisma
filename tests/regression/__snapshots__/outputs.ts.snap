// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputs should properly generate aggregate classes for model with lowercase name: AggregateExample 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { AggregateExampleCountArgs } from \\"./args/AggregateExampleCountArgs\\";

@TypeGraphQL.ObjectType({
  isAbstract: true,
  description: undefined,
})
export class AggregateExample {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false,
    description: undefined
  })
  count(@TypeGraphQL.Ctx() ctx: any, @TypeGraphQL.Args() args: AggregateExampleCountArgs) {
    return ctx.prisma.example.count(args);
  }
}
"
`;

exports[`outputs should properly generate aggregate classes for model with lowercase name: AggregateExampleCountArgs 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { ExampleOrderByInput } from \\"../../inputs/ExampleOrderByInput\\";
import { ExampleWhereInput } from \\"../../inputs/ExampleWhereInput\\";
import { ExampleWhereUniqueInput } from \\"../../inputs/ExampleWhereUniqueInput\\";

@TypeGraphQL.ArgsType()
export class AggregateExampleCountArgs {
  @TypeGraphQL.Field(_type => ExampleWhereInput, { nullable: true })
  where?: ExampleWhereInput | undefined;

  @TypeGraphQL.Field(_type => ExampleOrderByInput, { nullable: true })
  orderBy?: ExampleOrderByInput | undefined;

  @TypeGraphQL.Field(_type => ExampleWhereUniqueInput, { nullable: true })
  cursor?: ExampleWhereUniqueInput | undefined;

  @TypeGraphQL.Field(_type => TypeGraphQL.Int, { nullable: true })
  take?: number | undefined;

  @TypeGraphQL.Field(_type => TypeGraphQL.Int, { nullable: true })
  skip?: number | undefined;
}
"
`;

exports[`outputs should properly generate aggregate classes for model with lowercase name: outputs args index 1`] = `
"export { AggregateExampleCountArgs } from \\"./AggregateExampleCountArgs\\";
"
`;

exports[`outputs should properly generate aggregate classes for model with lowercase name: outputs index 1`] = `
"export { AggregateExample } from \\"./AggregateExample\\";
export { BatchPayload } from \\"./BatchPayload\\";
export * from \\"./args\\";
"
`;

exports[`outputs should properly generate aggregate classes for renamed model: AggregateExample 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { AggregateExampleCountArgs } from \\"./args/AggregateExampleCountArgs\\";

@TypeGraphQL.ObjectType({
  isAbstract: true,
  description: undefined,
})
export class AggregateExample {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false,
    description: undefined
  })
  count(@TypeGraphQL.Ctx() ctx: any, @TypeGraphQL.Args() args: AggregateExampleCountArgs) {
    return ctx.prisma.sample.count(args);
  }
}
"
`;

exports[`outputs should properly generate aggregate classes for renamed model: AggregateExampleCountArgs 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { ExampleOrderByInput } from \\"../../inputs/ExampleOrderByInput\\";
import { ExampleWhereInput } from \\"../../inputs/ExampleWhereInput\\";
import { ExampleWhereUniqueInput } from \\"../../inputs/ExampleWhereUniqueInput\\";

@TypeGraphQL.ArgsType()
export class AggregateExampleCountArgs {
  @TypeGraphQL.Field(_type => ExampleWhereInput, { nullable: true })
  where?: ExampleWhereInput | undefined;

  @TypeGraphQL.Field(_type => ExampleOrderByInput, { nullable: true })
  orderBy?: ExampleOrderByInput | undefined;

  @TypeGraphQL.Field(_type => ExampleWhereUniqueInput, { nullable: true })
  cursor?: ExampleWhereUniqueInput | undefined;

  @TypeGraphQL.Field(_type => TypeGraphQL.Int, { nullable: true })
  take?: number | undefined;

  @TypeGraphQL.Field(_type => TypeGraphQL.Int, { nullable: true })
  skip?: number | undefined;
}
"
`;

exports[`outputs should properly generate aggregate classes for renamed model: outputs args index 1`] = `
"export { AggregateExampleCountArgs } from \\"./AggregateExampleCountArgs\\";
"
`;

exports[`outputs should properly generate aggregate classes for renamed model: outputs index 1`] = `
"export { AggregateExample } from \\"./AggregateExample\\";
export { BatchPayload } from \\"./BatchPayload\\";
export * from \\"./args\\";
"
`;

exports[`outputs should properly generate args type classes for aggregate: AggregateSampleCountArgs 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { SampleOrderByInput } from \\"../../inputs/SampleOrderByInput\\";
import { SampleWhereInput } from \\"../../inputs/SampleWhereInput\\";
import { SampleWhereUniqueInput } from \\"../../inputs/SampleWhereUniqueInput\\";

@TypeGraphQL.ArgsType()
export class AggregateSampleCountArgs {
  @TypeGraphQL.Field(_type => SampleWhereInput, { nullable: true })
  where?: SampleWhereInput | undefined;

  @TypeGraphQL.Field(_type => SampleOrderByInput, { nullable: true })
  orderBy?: SampleOrderByInput | undefined;

  @TypeGraphQL.Field(_type => SampleWhereUniqueInput, { nullable: true })
  cursor?: SampleWhereUniqueInput | undefined;

  @TypeGraphQL.Field(_type => TypeGraphQL.Int, { nullable: true })
  take?: number | undefined;

  @TypeGraphQL.Field(_type => TypeGraphQL.Int, { nullable: true })
  skip?: number | undefined;
}
"
`;

exports[`outputs should properly generate args type classes for aggregate: outputs args index 1`] = `
"export { AggregateSampleCountArgs } from \\"./AggregateSampleCountArgs\\";
"
`;

exports[`outputs should properly generate output type classes: AggregateSample 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { AggregateSampleCountArgs } from \\"./args/AggregateSampleCountArgs\\";

@TypeGraphQL.ObjectType({
  isAbstract: true,
  description: undefined,
})
export class AggregateSample {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false,
    description: undefined
  })
  count(@TypeGraphQL.Ctx() ctx: any, @TypeGraphQL.Args() args: AggregateSampleCountArgs) {
    return ctx.prisma.sample.count(args);
  }
}
"
`;

exports[`outputs should properly generate output type classes: BatchPayload 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";

@TypeGraphQL.ObjectType({
  isAbstract: true,
  description: undefined,
})
export class BatchPayload {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false,
    description: undefined
  })
  count!: number;
}
"
`;

exports[`outputs should properly generate output type classes: outputs index 1`] = `
"export { AggregateSample } from \\"./AggregateSample\\";
export { BatchPayload } from \\"./BatchPayload\\";
export * from \\"./args\\";
"
`;
